packages/react-router-dom/__tests__/trailing-slashes-test.tsx

```diff
@@ -0,0 +1,454 @@
+                />
+              </Routes>
+            </BrowserRouter>,
+            node
+          );
+        });
+
+        expect(node.querySelectorAll("a")).toMatchInlineSnapshot(`
+          NodeList [
+            <a
+              href="/app/"
+            />,
+            <a
+              href="/app/"
+            />,
+          ]
+        `);
+      });
+
+      test('always contains trailing slashes on root links (path="" route)', () => {
+        let window = getWindowImpl("/app/");
+        act(() => {
+          ReactDOM.render(
+            <BrowserRouter basename="/app/" window={window}>
+              <Routes>
+                <Route
+                  path=""
+                  element={
+                    <>
+                      <Link to="" />
+                      <Link to="/" />
+                    </>
+                  }
+                />
+              </Routes>
+            </BrowserRouter>,
+            node
+          );
+        });
+
+        expect(node.querySelectorAll("a")).toMatchInlineSnapshot(`
+          NodeList [
+            <a
+              href="/app/"
+            />,
+            <a
+              href="/app/"
+            />,
+          ]
+        `);
+      });
+
+      test("always contains root trailing slashes in nested routes", () => {
+        let window = getWindowImpl("/app/parent/child");
+        act(() => {
+          ReactDOM.render(
+            <BrowserRouter basename="/app/" window={window}>
+              <Routes>
+                <Route
+                  path="parent"
+                  element={
+                    <>
+                      <Link to=".." />
+                      <Link to="../" />
+                      <Outlet />
+                    </>
+                  }
+                >
+                  <Route
+                    path="child"
+                    element={
+                      <>
+                        <Link to="../.." />
+                        <Link to="../../" />
+                        <Link to="../../parent" />
+                        <Link to="../../parent/" />
+                        <Outlet />
+                      </>
+                    }
+                  >
+                    <Route
+                      index
+                      element={
+                        <>
+                          <Link to="../../.." />
+                          <Link to="../../../" />
+                          <Link to="../../child" />
+                          <Link to="../../child/" />
+                        </>
+                      }
+                    />
+                  </Route>
+                </Route>
+              </Routes>
+            </BrowserRouter>,
+            node
+          );
+        });
+
+        expect(node.querySelectorAll("a")).toMatchInlineSnapshot(`
+          NodeList [
+            <a
+              href="/app/"
+            />,
+            <a
+              href="/app/"
+            />,
```

Comment:
When the `basename` has a trailing slash, routes to the root route will always have the trailing slash, but users still control trailing slash behavior on nested locations (see `/app/parent/child` below)

Reply:
Thank you for the review. `getWindowImpl` was missing part when I was trying to write some decent tests. Good to know how to manage such cases 👍